### PR TITLE
Ignore token-file argument when value not needed

### PR DESCRIPTION
**Pull request by kke**
## Description

Both "k0s controller" and "k0s worker" take either `[TOKEN]` as an argument or a path to a file containing the join token via `--token-file`.

Before this PR, when `--token-file` was given, the file was read upon spin-up, even when the node has joined already and won't utilize the token for anything. As the token is not needed after join, you would expect that you can or even should remove the file, but that makes k0s refuse to start unless the `--token-file` argument is also removed from the systemd unit or other startup scripts.

This PR makes the controller and worker read the contents of the file only right before the value is getting used. This allows `--token-file` argument to be ignored when it is not needed and the file not existing will not generate an error.

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [ ] Manual test
- [X] Auto test added

## Checklist:

- [X] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [X] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by kke** [comment]
Todo: some kind of test.

File: cmd/controller/controller.go [hunk 1]
```diff
@@ -177,8 +170,22 @@ func (c *command) start(ctx context.Context) error {
 
 	var joinClient *token.JoinClient
 
-	if c.TokenArg != "" && c.needToJoin() {
-		joinClient, err = joinController(ctx, c.TokenArg, c.K0sVars.CertRootDir)
+	if c.needToJoin() {
+		if c.TokenArg == "" && c.TokenFile == "" {
+			return fmt.Errorf("token argument or --token-file is required to join")
+		}
```
Comment (twz123) [inline cmd/controller/controller.go:176]: If we want to keep the same behavior as before it, it needs to be

```suggestion
	if (c.TokenArg != "" || c.TokenFile != "") && c.needToJoin() {
```

Comment (kke) [reply cmd/controller/controller.go:176]: But is that a wished behavior? :)

**Comment by twz123** [comment]
This might need #4288.

File: inttest/upgrade/upgrade_test.go [hunk 2]
```diff
@@ -151,6 +151,11 @@ func (s *UpgradeSuite) TestK0sGetsUp() {
 			if err != nil {
 				return err
 			}
+			// delete token file because it shouldn't be needed anymore to see that worker can join without it
```
Comment (twz123) [inline inttest/upgrade/upgrade_test.go:154]:
```suggestion
			// Delete the token file, as it shouldn't be needed after the worker has joined.
```

File: inttest/hacontrolplane/hacontrolplane_test.go [hunk 3]
```diff
@@ -108,7 +109,9 @@ func (s *HAControlplaneSuite) TestDeregistration() {
 	defer sshC1.Disconnect()
 	_, err = sshC1.ExecWithOutput(s.Context(), "kill $(pidof k0s) && while pidof k0s; do sleep 0.1s; done")
 	s.Require().NoError(err)
-	s.NoError(s.InitController(1, token))
+	_, err = sshC1.ExecWithOutput(s.Context(), "rm -f /etc/k0s.token")
```
Comment (twz123) [inline inttest/hacontrolplane/hacontrolplane_test.go:112]:
```suggestion
	// Delete the token file, as it shouldn't be needed after the controller has joined.
	_, err = sshC1.ExecWithOutput(s.Context(), "rm -f /etc/k0s.token")
```

**Review by twz123 — APPROVED** [review]

**Review by twz123 — APPROVED** [review]

**Pull request merged**